stage0/src/fw_cfg.rs

```diff
@@ -0,0 +1,203 @@
+        let mut fwcfg = Self {
+            selector: port_factory.new_writer(FWCFG_PORT_SELECTOR),
+            data: port_factory.new_reader(FWCFG_PORT_DATA),
+        };
+
+        // Make sure the fw_cfg device is available. If the device is not available, writing and
+        // reading to I/O ports is undefined behaviour.
+        fwcfg.write_selector(FwCfgItems::Signature as u16)?;
+        let mut signature = [0u8; SIGNATURE.len()];
+        fwcfg.read(&mut signature)?;
+
+        if signature == SIGNATURE {
+            Ok(fwcfg)
+        } else {
+            Err("QEMU fw_cfg device not available")
+        }
+    }
+
+    /// Returns an iterator over the files in the fw_cfg system.
+    ///
+    /// # Safety
+    ///
+    /// You should consume the iterator fully before calling any other methods of this struct, or at
+    /// the very least don't use the iterator after calling some other function.
+    /// (This means that, for example, you shouldn't try to read the files while iterating over the
+    /// directory contents.)
+    /// If you call any other methods, the iterator will be in an undefined state and unsafe to read
+    /// from.
+    pub unsafe fn dir(&mut self) -> impl Iterator<Item = DirEntry> + '_ {
+        self.write_selector(FwCfgItems::FileDir as u16).unwrap();
+
+        // We don't represent FwCfgFiles as a struct, as you can't do that in safe Rust. Therefore,
+        // read fields individually.
+        let count = {
+            let mut buf = [0u8; 4];
+            self.read(&mut buf).unwrap();
+            u32::from_be_bytes(buf)
+        };
+
+        (0..count).map(|_| {
+            let mut file = DirEntry::default();
+            self.read(&mut file).unwrap();
+            file
+        })
+    }
+
+    /// Reads the contents of a file to a predetermined struct.
+    /// Returns the number of bytes read.
+    ///
+    /// # Safety
+    ///
+    /// While we will never write beyond the memory allocated to `object`, it is up to the caller to
+    /// ensure that the type of `object` is a faithful representation of the contents of the file;
+    /// otherwise, `object` may be left in an invalid state.
+    pub unsafe fn read_file_by_name<T>(
+        &mut self,
+        name: &CStr,
+        object: &mut T,
+    ) -> Result<usize, &'static str> {
+        // Safety: we're not using the iterator after we've found the file.
+        let mut entry: Option<DirEntry> = None;
+        for file in unsafe { self.dir() } {
+            if file.name() != name {
+                continue;
+            };
+            entry = Some(file);
+            break;
+        }
+        if let Some(file) = entry {
+            // Safety: we dereference a valid pointer (as we get it from a reference) and the
+            // resulting array will have the same size as the underlying type.
+            let buf = unsafe {
+                core::slice::from_raw_parts_mut(
```

Comment:
This bit of unsafe code (and below) can be made a bit safer by deriving AsBytes and FromBytes from the `zerocopy` crate. `AsBytes` basically does the same as this, but deriving `FromBytes` ensures that the object will not cause undefined behaviour when writing arbitrary bytes to the memory by e.g. checking alignment and ensuring that none of the fields make assumptions about the underlying bytes.

https://docs.rs/zerocopy/latest/zerocopy/trait.AsBytes.html#method.as_bytes_mut

Reply:
Cool, I wasn't familiar with the  `zerocopy` crate!